File: backend/src/main/java/com/NBE_4_5_2/Team5/domain/chat/entity/ChatMessage.java
package com.NBE_4_5_2.Team5.domain.chat.entity;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

import jakarta.persistence.*;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor
@Entity
public class ChatMessage {
	@Id
	private String messageId = UUID.randomUUID().toString();

	private MessageType type; // 메시지 타입
	private String roomId; // 방번호
	private String sender; // 메시지 보낸사람
	private String receiver;
	private String message; // 메시지
	private String image;
	private long userCount; // 채팅방 인원수, 채팅방 내에서 메시지가 전달될때 인원수 갱신시 사용
	private String timestamp;
	private Float latitude = 0.0f;
	private Float longitude = 0.0f;

	@ElementCollection
	private Map<String, Boolean> deleteStatus = new HashMap<>();

	@Builder
	public ChatMessage(MessageType type, String roomId, String sender, String receiver, String message, String image,
		long userCount, Float latitude, Float longitude) {
		messageId = UUID.randomUUID().toString();
		this.type = type;
		this.roomId = roomId;
		this.sender = sender;
		this.receiver = receiver;
		this.message = message;
		this.image = image;
		this.userCount = userCount;
		this.timestamp = formatTimestamp(LocalDateTime.now());
		this.latitude = latitude;
		this.longitude = longitude;

		deleteStatus.put(sender, false);
		deleteStatus.put(receiver, false);
	}

	// 메시지 타입 : 입장, 퇴장, 채팅, 이미지 추가
	public enum MessageType {
		ENTER, QUIT, TALK, IMAGE, LOCATION
	}

	public String formatTimestamp(LocalDateTime timestamp) {
		DateTimeFormatter formatter = DateTimeFormatter.ofPattern("HH:mm");
		return timestamp.format(formatter);
	}

	public void setDeleteStatus(String username,boolean status) {
		this.deleteStatus.put(username,status);
	}

	public Boolean getDeleteStatus(String username) {
		return this.deleteStatus.get(username);
	}

}

File: backend/src/main/java/com/NBE_4_5_2/Team5/domain/chat/entity/ChatMessage.kt
package com.NBE_4_5_2.Team5.domain.chat.entity

import com.fasterxml.jackson.annotation.JsonProperty
import jakarta.persistence.ElementCollection
import jakarta.persistence.Entity
import jakarta.persistence.Id
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*

@Entity
class ChatMessage (
    private var _type: MessageType = MessageType.TALK,
    private var _roomId: String="",
    private var _sender: String="",
    private val _receiver: String="",
    private var _message: String="",
    @JsonProperty("image")
    private var _image: String="",
    private var _userCount: Long=0,
    @JsonProperty("latitude")
    private var _latitude: Float=0.0f,
    @JsonProperty("longitude")
    private var _longitude: Float=0.0f
) {
    @Id
    private val _messageId = UUID.randomUUID().toString()
    private var _timestamp: String = formatTimestamp(LocalDateTime.now())

    @ElementCollection
    private val deleteStatus: MutableMap<String, Boolean> = HashMap()

    fun getMessageId(): String {
        return _messageId
    }
    fun getType(): MessageType {
        return _type
    }
    fun setType(type: MessageType) {
        _type=type
    }
    fun getRoomId(): String {
        return _roomId
    }
    fun setRoomId(value: String) {
        _roomId=value
    }
    fun getSender(): String {
        return _sender
    }
    fun setSender(sender: String) {
        _sender=sender
    }
    fun getReceiver(): String {
        return _receiver
    }
    fun getMessage(): String {
        return _message
    }
    fun setMessage(message: String) {
        _message = message
    }
    fun getTimestamp(): String {
        return _timestamp
    }
    fun getImage(): String {
        return _image
    }
    fun getUserCount(): Long {
        return _userCount
    }
    fun setUserCount(userCount: Long) {
        _userCount=userCount
    }
    fun getLatitude(): Float {
        return _latitude
    }
    fun getLongitude(): Float {
        return _longitude
    }

    init {
        deleteStatus[_sender] = false
        deleteStatus[_receiver] = false
    }

    // 메시지 타입 : 입장, 퇴장, 채팅, 이미지 추가
    enum class MessageType {
        ENTER, QUIT, TALK, IMAGE, LOCATION
    }

    companion object{
        fun formatTimestamp(timestamp: LocalDateTime): String {
            val formatter = DateTimeFormatter.ofPattern("HH:mm")
            return timestamp.format(formatter)
        }
    }

    fun setDeleteStatus(username: String, status: Boolean) {
        deleteStatus[username] = status
    }

    fun getDeleteStatus(username: String?): Boolean? {
        return deleteStatus[username]
    }
    override fun toString(): String {
        return "ChatMessage(messageId='$_messageId', type='$_type', roomId='$_roomId', sender='$_sender', receiver='$_receiver', message='$_message', image='$_image', userCount=$_userCount, latitude=$_latitude, longitude=$_longitude, timestamp='$_timestamp')"
    }
}